apps/backend/src/libs/plugins/authorization/authorization.plugin.ts

```diff
@@ -0,0 +1,69 @@
+import { type FastifyRequest } from "fastify";
+import fp from "fastify-plugin";
+import { JWTExpired } from "jose/errors";
+
+import { ExceptionMessage } from "~/libs/enums/enums.js";
+import { HTTPCode } from "~/libs/modules/http/http.js";
+import { type Token } from "~/libs/modules/token/token.js";
+import { AuthError } from "~/modules/auth/auth.js";
+import { type UserService } from "~/modules/users/user.service.js";
+
+type Options = {
+	token: Token;
+	userService: UserService;
+	whiteRoutes: readonly RegExp[];
+};
+
+const authorization = fp<Options>((fastify, options, done) => {
+	const { token, userService, whiteRoutes } = options;
+
+	fastify.decorateRequest("user", null);
+
+	fastify.addHook("preHandler", async (request: FastifyRequest) => {
+		const { url } = request.raw;
+
+		for (const whiteRoute of whiteRoutes) {
+			if (whiteRoute.test(url as string)) {
+				return;
+			}
+		}
+
+		const jwtToken = request.headers["authorization"];
+
+		if (!jwtToken) {
+			throw new AuthError({
+				message: ExceptionMessage.NO_TOKEN_PROVIDED,
+				status: HTTPCode.UNAUTHORIZED,
+			});
+		}
+
+		try {
+			const payload = await token.verifyToken(jwtToken);
+			const { userId } = payload;
+
+			if (userId) {
```

Comment:
We need to throw error if there is no userId

Reply:
if there is no userId in payload, the result is already "Invalid token." error.
should it throw smth else?

Reply:
Can't there be a case when the token is valid but contains any other info like project_id? if we have different tokens in the app for example? Or actually I'm thinking it still won't find user in this case 🤔 then we don't need this if condition to check if userId is present